cmd/promtool/unittest.go

```diff
@@ -323,11 +325,47 @@ func (tg *testGroup) test(evalInterval time.Duration, groupOrderMap map[string]i
 					}
 					expString := indentLines(expAlerts.String(), "            ")
 					gotString := indentLines(gotAlerts.String(), "            ")
-					errs = append(errs, fmt.Errorf("%s    alertname: %s, time: %s, \n        exp:%v, \n        got:%v",
-						testName, testcase.Alertname, testcase.EvalTime.String(), expString, gotString))
+
+					// If empty, populates an empty value
+					if gotAlerts.Len() == 0 {
+						gotAlerts = append(gotAlerts, labelAndAnnotation{
+							Labels:      labels.Labels{},
+							Annotations: labels.Labels{},
+						})
+					}
+					// If empty, populates an empty value
+					if expAlerts.Len() == 0 {
+						expAlerts = append(expAlerts, labelAndAnnotation{
+							Labels:      labels.Labels{},
+							Annotations: labels.Labels{},
+						})
+					}
+
+					diffOpts := jsondiff.DefaultConsoleOptions()
+					expAlertsJSON, err := json.Marshal(expAlerts)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("alertName: %s\nError converting expAlert to JSON: %v", tg.TestGroupName, expAlerts.String()))
+						continue
+					}
+
+					gotAlertsJSON, err := json.Marshal(gotAlerts)
+					if err != nil {
+						errs = append(errs, fmt.Errorf("alertName: %s\nError converting gotAlert to JSON: %v", tg.TestGroupName, gotAlerts.String()))
+						continue
+					}
+
+					res, diff := jsondiff.Compare(expAlertsJSON, gotAlertsJSON, &diffOpts)
+					if res != jsondiff.FullMatch {
+						if runtime.GOOS == "windows" {
+							errs = append(errs, fmt.Errorf("%s    alertname: %s, time: %s, \n        exp:%v, \n        got:%v",
+								testName, testcase.Alertname, testcase.EvalTime.String(), expString, gotString))
+						} else {
+							errs = append(errs, fmt.Errorf("%s    alertname: %s, time: %s, \n        diff:%v",
```

Comment:
I'm also not keen on just showing "diff", it's not clear to the user which way around the diff is. Is red got and green exp? We probably at least want a way to show exp+got+diff (a --diff option?).

We should also support something that isn't color based, particularly if --color=auto checks stdout is a terminal, but also because many CI systems won't support ANSI colours either and the output would be quite useless then.

Reply:
The colors are other way around, the `red` is something we want but that's gone missing. The `green` is something we don't want but came as extra. We can swap the colors, if required.